lib/system/io.nim

```diff
@@ -792,20 +792,21 @@ when declared(stdout):
         proc flockfile(f: File) {.importc, nodecl.}
         proc funlockfile(f: File) {.importc, nodecl.}
         flockfile(stdout)
+        defer: funlockfile(stdout)
       when defined(windows) and compileOption("threads"):
         acquireSys echoLock
+        defer: releaseSys echoLock
       for s in args:
         when defined(windows):
           writeWindows(stdout, s)
         else:
-          discard c_fwrite(s.cstring, cast[csize_t](s.len), 1, stdout)
+          if c_fwrite(s.cstring, cast[csize_t](s.len), 1, stdout) != s.len:
```

Comment:
CI logs for windows show https://dev.azure.com/nim-lang/Nim/_build/results?buildId=11013&view=logs&jobId=9ce34ca5-78e3-50eb-e582-5eabb51cd3ed&j=9ce34ca5-78e3-50eb-e582-5eabb51cd3ed&t=96a0fe61-4831-51a3-7585-5f390132c17a

2020-12-16T17:55:31.7097151Z Error: unhandled exception: expected raising 'IOError', instead raised 'IndexDefect' by: 

I think that's because of this:
```nim
when defined(windows):
  proc writeWindows(f: File; s: string; doRaise = false) =
    # Don't ask why but the 'printf' family of function is the only thing
    # that writes utf-8 strings reliably on Windows. At least on my Win 10
    # machine. We also enable `setConsoleOutputCP(65001)` now by default.
    # But we cannot call printf directly as the string might contain \0.
    # So we have to loop over all the sections separated by potential \0s.
    var i = c_fprintf(f, "%s", s)
    while i < s.len:
      ...
```

which sounds super buggy:
```
EXIT STATUS
     The printf utility exits 0 on success, and >0 if an error occurs.

```

so the fix seems to be to fix this line:

```
while i < s.len:
```

Reply:
I think `printf` is different than `fprintf`. It looks like `fprintf` *does* return the number of bytes written: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fprintf-fprintf-l-fwprintf-fwprintf-l?view=msvc-160

Except in the error case, it returns -1, so I think that's the problem.

Reply:
my bad I hallucinated printf for fprintf, but ya that code was stll buggy regardless :), and your latest commit fixes it :)